### ISSUE TITLE
VS memory leak opening solutions with roslyn analysers

### ISSUE THREAD

**Issue by mjcheetham** [dotnet/roslyn#10655]
Hi Roslyn team,
I believe I've found a memory leak and performance issue surrounding Roslyn analysers after having opened and closed several solutions in the same Visual Studio session (w/o closing `devenv.exe` in between).

The issue occurs when you have a set of analysers installed via a VSIX, and proceed to open, close and re-open solutions. I've been able to push the memory usage of VS up arbitrarily high by continuing to switch between solutions. I've exhibited the issue using a few different VSIX extensions
which leads me to believe this is not specific to a particular set of analysers.

I've captured memory dumps of VS in three cases: two with analyser VSIXs installed and one without, and also done some quick analysis of them to try and narrow down the offender.

From these dumps it looks like there's a problem with some task continuations regarding a `DiagnosticsUpdated` event, presumably trigged when the set of diagnostic analysers have changed (at solution open/close). As you'll see below, there are several million thread/task objects all chained together holding onto each provided analyser.

All dumps were created of VS2015 U2 using the 32-bit `taskmgr.exe` 'Create dump file' on a 64-bit Windows 10 machine. I was using WinDBG to look at the heaps. I also tried to use dotMemory to compare snapshots before and after opening solutions and also found the same thread context/task
continuation object has been created (see below).

Thanks,
Matthew

My overview of the dumps can be found [here](https://gist.github.com/mjcheetham/4f5d01b94e5231ccf929b2fff2b294f9).
The actual dump files are rather large at 1-2 GB each so please contact me if you would like a copy of them :smile:

<hr>

**Version Used**: VS2015 U2

**Steps to Reproduce**:

These steps are the same for all dumps, the only difference is which (if any) extension was installed.
Sample solutions can be found [here](https://github.com/mjcheetham/MegaMix), [here](https://github.com/SonarSource-VisualStudio/sonarlint-visualstudio) and [here](https://github.com/SonarSource-VisualStudio/sonar-msbuild-runner).

1. Open task manager (32-bit version if you want to take a dump of the process) and note the initial memory usage (about 100-300 MB on my install)
1. Open Visual Studio with SonarLint 2.1 installed
1. Open a sample solution containing managed code
1. Wait for the solution to load enough so that the error list starts to be populated
1. Repeat and open the other another sample solutions, waiting a short while for VS to initialise the projects
1. Keep switching back and forth between these solutions in quick succession (as many times as desired).
To make reproing this easier, open the Start Page (View > Start Page) and check the
'Keep page open after project loads' option, so you can just mash the mouse-click on the recent solutions links.
1. Close the solution and force a GC (default VS chord <kbd>Ctrl+Shift+Alt+F12</kbd> twice)
1. Note how the memory usage has dropped, but is still significantly higher that it should be (I managed to get it to approx. 2GB).

**NB:** Not too sure about whether or not switching between solutions before they've fully finished
loading is a requirement for this repro. I do find that if you switch before any of the managed projects are loaded at all then no analysers are loaded (and then no issue presents itself).

**Expected Behavior**:
Memory is released after closing a solution.

**Actual Behavior**:
Memory continues to be consumed.

**Comment by heejaechang** [comment]
if you give VS to chance to process tasks, does it goes away?

**Comment by heejaechang** [comment]
what I think is going on is that when solution is closed/opened, we are enqueuing async events to clean things up and you are opening closing solution before they are all processed making all those async events are accumulated which should eventually all processed.

so question is it is ever all processed. if not it is leak, otherwise it is how things work in U2. we are changing it in U3.

**Comment by heejaechang** [comment]
the reason we use async events rather than synchronous events is so that we don't block user from closing solution. otherwise, we need to block solution close which we don't want to do.

issue created - https://github.com/dotnet/roslyn/issues/10703

**Comment by heejaechang** [comment]
does sonarlint install some per project analyzer? if it does, then we do create one instance per project.

**Comment by mavasani** [comment]
@heejaechang Did you verify from the repro that the leak is due to the async events being enqueued? Otherwise, I can run the repro and confirm the source of the leak.

**Comment by heejaechang** [comment]
if you go to link you shared. https://gist.github.com/mjcheetham/4f5d01b94e5231ccf929b2fff2b294f9

it shows the async event issue.

**Comment by mavasani** [comment]
Thanks, I am moving over this bug to you then.

**Comment by mjcheetham** [comment]
@heejaechang 
> if you give VS to chance to process tasks, does it goes away?

I hadn't noticed/tried explicitly at the time.
After trying now... pushing it up to 800/900 MB and then leaving VS alone (with no solutions open), it continued to consume more memory(!) (but hardly any CPU - 0.3%). It wasn't until I forced a GC after waiting nearly 4 minutes after closing the last solution.
Even then it was about 2x above the initial launch baseline usage (but perhaps this is to be expected?)

See memory usage graph attached
1. VS open
2. Repo steps start (open + close solutions)
3. Close final solution
4. Forcing a GC myself

![roslyn_10655](https://cloud.githubusercontent.com/assets/5658207/14669101/4ed2e166-06df-11e6-99ea-14af6a21df3e.png)

**Comment by mjcheetham** [comment]
@heejaechang 

> does sonarlint install some per project analyzer? if it does, then we do create one instance per project.

One/two of the test solutions I was opening also contained Roslyn analyser NuGet packages (SonarAnalyzer and Microsoft.CodeAnalysis.Analyzers) on most projects. So this could explain the 30 instances of each rule.

**Comment by heejaechang** [comment]
@mjcheetham thank you for taking a time to provide more information.

I also spent some time to look into this. and found 2 leaks we have. but that is only part of it, async events being accumulated is something by design. we don't expect users to continuously open and close solution in quick succession. in normal usage, async event won't be accumulated as much.

but also that being said, our recent change (https://github.com/dotnet/roslyn/pull/10509) reduces number of async events generated. when I tested with my leak fix, async event didn't accumulated as fast as it was before. memory mostly stayed same.
